### DISCUSSION TITLE
3D Transformation

### DISCUSSION THREAD

**Discussion by VedankNaik** [iTwin/itwinjs-core#3569] [Q&A]
Following [Simple 3D](https://www.itwinjs.org/sandboxes/iTwinPlatform/Simple%203d) sample. 
I need to perform transformations on shapes. How should I go about this?
Also considering there are multiple shapes in view how can I select particular shape and perform transformations on selected shape.

Thanks.

**Comment by bbastings** [comment]
That sample demonstrates a blank connection that creates some view decorations, the geometry is not persistent...is that really what you have and you are not looking to apply a transform to persistent elements?

**Reply by VedankNaik** [reply]
I will be using imodel where I need to place primitive shapes, sample has shapes similar to what I'm looking for. Above sample looks closest to what I'm looking for so I've been following the same. Is there other way to implement this?
Also we might have to save those shapes in any way possible. 

Thanks.

**Reply by bbastings** [reply]
I can't tell if you want to create an interactive editing application that inserts new geometric elements (shapes) and later modifies them to update their placement transform or you are just looking for a way to make view decorations like the Simple 3D sample pickable?

**Reply by VedankNaik** [reply]
I want create an interactive editing application to insert shapes and modify them later.

**Comment by bbastings** [comment]
You can have a look at the tools in the [editor](https://github.com/iTwin/itwinjs-core/tree/master/editor) package.

You can find examples of placement tools in [SketchTools.ts](https://github.com/iTwin/itwinjs-core/blob/master/editor/frontend/src/SketchTools.ts).
You can find examples of modification tools in [TransformElementsTool.ts](https://github.com/iTwin/itwinjs-core/blob/master/editor/frontend/src/TransformElementsTool.ts).

Be aware that the editing apis are still in alpha, but you are welcome to try them out and provide feedback.

**Reply by VedankNaik** [reply]
Thanks. I'll check this out. 
Can you tell me how should I go about this?

**Comment by bbastings** [comment]
Sorry for the delay, didn't notice the edit.

What's your starting point? Do you have an app you've written that can open an imodel for write?

**Reply by VedankNaik** [reply]
Yes, I've setup app with imodel. By write does it mean setting up imodel access from imodelhub? If so then yes I've set that too.

**Reply by pmconne** [reply]
Editing the contents of an iModel requires obtaining a [briefcase](https://www.itwinjs.org/learning/frontend/imodelconnection/) that can communicate with a dedicated backend - typically running on the same machine as the frontend - and push changes to the hub.
We are currently working on an example authoring app that can set up a briefcase and make use of the editing tools @bbastings previously linked.
Until then if you are comfortable experimenting with `@alpha` features you can glean some information from the editing capabilities built into [ui-test-app](https://github.com/iTwin/itwinjs-core/tree/master/test-apps/ui-test-app).

**Comment by VedankNaik** [comment]
Hello @pmconne @bbastings,

Will [PrimitiveTool](https://www.itwinjs.org/reference/core-frontend/tools/primitivetool/) help us in this? It mentions can be used to implement tools to create or modify geometric elements. 
Also we did try to implement it from snippets in [Writing a PrimitiveTool](https://www.itwinjs.org/learning/frontend/primitivetools/) but it returns us false for `isCompatibleViewport`. Does this need any prior setup for tools to work?

Thank you.

**Reply by pmconne** [reply]
> Will [PrimitiveTool](https://www.itwinjs.org/reference/core-frontend/tools/primitivetool/) help us in this? It mentions can be used to implement tools to create or modify geometric elements.

That is its primary purpose but you can also use it to create transient geometry.
`isCompatibleViewport` is returning false because `PrimitiveTool.requireWriteableTarget` returns true by default, assuming that you want to create elements, not transient geometry. Override it to return false if you want transient geometry. If you want to create elements, you need a writeable connection to an iModel - see previous discussion above.

**Reply by VedankNaik** [reply]
Can you tell me about opening BriefcaseConnection? So far we have been using CheckpointConnection with iTwinId, iModelId.

**Reply by pmconne** [reply]
You can learn about BriefcaseConnection by reading [the documentation](https://www.itwinjs.org/learning/frontend/imodelconnection/).
As @bbastings already said, the editing APIs are currently `alpha` and you should only be trying to use them if you are **comfortable experimenting** with largely undocumented features.
If you want to edit an iModel you probably should be doing so using a desktop - not browser-based - app.
You can search around the [ui-test-app](https://github.com/iTwin/itwinjs-core/tree/master/test-apps/ui-test-app) code to see how it handles editing.

**Reply by VedankNaik** [reply]
> Override it to return false if you want transient geometry. 

Tried overriding and was able to run drawing line tool provided in [Writing a PrimitiveTool](https://www.itwinjs.org/learning/frontend/primitivetools/)
How can I move forward with creating other geometries? What should I follow?

**Reply by bbastings** [reply]
What specific types of geometry would you like to create?

You can create a shape by adding a LineString3d (sub-class of CurvePrimitive) to a Loop.
You can create various 3D solids and surfaces, Cone, Box, TorusPipe, etc. by creating the appropriate SolidPrimitive sub-class.

Look at the create methods for the geometry class you wish to create to see the inputs your PrimitiveTool needs to provide.

**Reply by VedankNaik** [reply]
Yes I'm looking for 3D solids (Cone, Box, TorusPipe, etc.).  Are there any samples which I can refer to?

Thanks.

**Comment by pmconne** [comment]
> Yes I'm looking for 3D solids (Cone, Box, TorusPipe, etc.). Are there any samples which I can refer to?

[Yes](https://developer.bentley.com/samples/).
![image](https://user-images.githubusercontent.com/22944042/174781004-6f26f42b-0836-4afe-b9e1-ac931d024bdc.png)

[In particular](https://www.itwinjs.org/sandboxes/iTwinPlatform/Simple%203d).
![image](https://user-images.githubusercontent.com/22944042/174781700-bb11fd5a-f491-48b1-891c-c6eb134f5bbd.png)

**Reply by VedankNaik** [reply]
I've already tried simple 3d sample. How do I create or modify elements with tools? Also once I get solids following the 3d sample how can I perform transformations like rotation, translation etc on those shapes?

**Reply by pmconne** [reply]
The base [GeometryQuery](https://www.itwinjs.org/reference/core-geometry/curve/geometryquery/) class from which the solid primitive types derive provides methods for applying [Transforms](https://www.itwinjs.org/reference/core-geometry/cartesiangeometry/transform/). Please peruse this and related documentation.

We have provided you with examples of (1) how to write a primitive tool to create geometry and (2) how to create solid geometry. What more - specifically - do you need to know to combine these to create your own tool?

**Reply by VedankNaik** [reply]
Can you clear my doubts? What can tools do exactly with geometries? 
Above 3D sample shows creating solids with decorators, what can tools do any different from this?

**Reply by bbastings** [reply]
> Can you clear my doubts? What can tools do exactly with geometries?

Tools can create new elements or modify existing elements.

The example tools for translating and rotating sets of elements was mentioned previously [TransformElementsTool.ts](https://github.com/iTwin/itwinjs-core/blob/master/editor/frontend/src/TransformElementsTool.ts). NOTE: These tools do not modify the element's GeometryStream, they instead apply a transform to the element's placement.

These example tools do demonstrate modifying the GeometryStream of an existing element [ModifyCurveTools.ts](https://github.com/iTwin/itwinjs-core/blob/master/editor/frontend/src/ModifyCurveTools.ts).

Additional example tools that create spheres, cylinders, and cones were added recently [SolidPrimitiveTools.ts](https://github.com/iTwin/itwinjs-core/blob/master/editor/frontend/src/SolidPrimitiveTools.ts).

Does that help you?

**Comment by VedankNaik** [comment]
Hello, 

I tried using editor package and tried some tools too. I am not using briefcase currently and since editing requires briefcase for trial purpose I tried overriding `isBriefcaseConnection` to return true.

For example I tried `CreateSphereTool` and I could see this options (image attached)
![image](https://user-images.githubusercontent.com/47364632/177511330-cbd95a30-e093-4462-99f1-bb238c387a0c.png)
But couldn't get sphere in the viewer. Is it because it requires a briefcase?

**Reply by pmconne** [reply]
> for trial purpose I tried overriding isBriefcaseConnection to return true

That's not how you get a briefcase.
It sounds to me like you should wait until support for authoring is out of alpha.
